test/test_module_init.py

```diff
@@ -0,0 +1,330 @@
+    if device is None:
+        args = [instantiate_class(arg, *get_example_args(arg, constructor_arg_db))
+                if inspect.isclass(arg) else arg for arg in args]
+        kwargs = {k: instantiate_class(v, *get_example_args(v, constructor_arg_db))
+                  if inspect.isclass(v) else v for k, v in kwargs.items()}
+    else:
+        args = [instantiate_class(arg, *get_example_args(arg, constructor_arg_db), device=device)
+                if inspect.isclass(arg) else arg for arg in args]
+        kwargs = {k: instantiate_class(v, *get_example_args(v, constructor_arg_db), device=device)
+                  if inspect.isclass(v) else v for k, v in kwargs.items()}
+        kwargs['device'] = device
+    return args, kwargs
+
+
+def generate_tests(test_cls, constructor_arg_db):
+    # test all modules underneath these namespaces...
+    NAMESPACES = [
+        torch.nn,
+        torch.nn.qat,
+        torch.nn.quantizable,
+    ]  # TODO: Add torch.nn.quantized
+    # ...except these
+    MODULES_TO_SKIP = {
+        torch.nn.Module,
+        torch.nn.Container,  # deprecated
+        torch.nn.NLLLoss2d,  # deprecated
+    }
+    # no need to support kwargs for these modules even though
+    # they have parameters / buffers because they are passed in
+    # already instantiated
+    MODULES_WITHOUT_KWARGS_SUPPORT = {
+        torch.nn.BCELoss,
+        torch.nn.BCEWithLogitsLoss,
+        torch.nn.CrossEntropyLoss,
+        torch.nn.FractionalMaxPool2d,
+        torch.nn.FractionalMaxPool3d,
+        torch.nn.MultiLabelSoftMarginLoss,
+        torch.nn.MultiMarginLoss,
+        torch.nn.NLLLoss,
+        torch.nn.TransformerDecoder,
+        torch.nn.TransformerEncoder,
+    }
+    # modules that supported kwargs before
+    MODULES_WITH_PREVIOUS_KWARGS = {
+        torch.nn.Identity,
+    }
+    # lazy modules don't instantiate parameters right away
+    LAZY_MODULES = {
+        torch.nn.LazyBatchNorm1d,
+        torch.nn.LazyBatchNorm2d,
+        torch.nn.LazyBatchNorm3d,
+        torch.nn.LazyConv1d,
+        torch.nn.LazyConv2d,
+        torch.nn.LazyConv3d,
+        torch.nn.LazyConvTranspose1d,
+        torch.nn.LazyConvTranspose2d,
+        torch.nn.LazyConvTranspose3d,
+        torch.nn.LazyConvTranspose3d,
+        torch.nn.LazyLinear,
+    }
+
+    for namespace in NAMESPACES:
+        # the "nn" in "torch.nn"
+        namespace_basename = namespace.__name__.split('.')[-1]
+        # namespace object (e.g. torch.nn.modules)
+        mod_namespace = getattr(namespace, 'modules')
+        # all module names (e.g. "Linear") in the namespace
+        module_names = getattr(mod_namespace, '__all__')
+        for module_name in module_names:
+            # class object for this module (e.g. torch.nn.Linear)
+            module_cls = getattr(mod_namespace, module_name)
+            if module_cls in MODULES_TO_SKIP: continue
+
+            # Create a function to run the test and setattr it onto the test class.
```

Comment:
I'm not saying this should be the PR that creates ModuleInfo... but MAYBE this should be the PR that does it. In particular, generators like this that are lengthy and complicated was one of the reasons we wanted to switch to decorating test templates that look like more typical tests.

Reply:
I for one am excited for `ModuleInfo`! As discussed offline, we'll meet sometime shortly after this PR to design what that could look like.

Reply:
Also adding a TODO in the code to refactor the constructor arg DB into `ModuleInfo`.